Adding PhasedQubitSparsePauli and PhasedQubitSparsePauliList

DanPuzzuoli opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This PR adds `PhasedQubitSparsePauli` and `PhasedQubitSparsePauliList`. These classes are very similar in motivation/functionality as `QubitSparsePauli` and `QubitSparsePauliList` (which are phase-less), but for the case where we want to store the phases of the Paulis.

The phases in `PhasedQubitSparsePauli` and `PhasedQubitSparsePauliList` will be stored in the same format as `Pauli`:
- Internally the ZX-phase is stored, but the phase attribute will return the "group phase". E.g. `Pauli("Y")._phase == 1` (as Y = ((-i)^1)ZX, whereas `Pauli("Y").phase == 0`.

### Details and comments

This PR is pretty straightforward; the interfaces for these class are pretty much the same as the phase-less version, but now everything needs to be updated to include a phase (or a list of phases). I've implemented `PhasedQubitSparsePauli` to store a `QubitSparsePauli` and a phase, and `PhasedQubitSparsePauliList` to store a `QubitSparsePauliList` and a vector of phases.

One bit of awkwardness imo is that you can't include phases in the `PhasedQubitSparsePauliList.from_list` and `PhasedQubitSparsePauli.from_label` constructors. I tried to add parsing of the strings in rust but got frustrated and moved on; perhaps this can be discussed/added in the review process.

DanPuzzuoli committed:
attempting to add initial scaffolding

DanPuzzuoli committed:
first successful build

DanPuzzuoli committed:
getting correct phase storage

DanPuzzuoli committed:
adding the phase to PhasedQubitSparsePauli

DanPuzzuoli committed:
initial addition of compose

DanPuzzuoli committed:
refactoring PyQubitSparsePauli.from_label

DanPuzzuoli committed:
adding indices and paulis attributes

DanPuzzuoli committed:
adding back tests

DanPuzzuoli committed:
adding eq based on modular phase and testing compose

DanPuzzuoli committed:
readding commutes tests

DanPuzzuoli committed:
adding from_sparse_label and tests

DanPuzzuoli committed:
re-adding sparse label construction

DanPuzzuoli committed:
changing from_raw_parts to have a default phase of 0 and interpret phase as group phase

DanPuzzuoli committed:
reintroducing all tests for PhasedQubitSparsePauli

DanPuzzuoli committed:
Completed pass on PhasedQubitSparsePauli docs

DanPuzzuoli committed:
repr appears to be working for now

DanPuzzuoli committed:
fixing eq function

DanPuzzuoli committed:
adding from_sparse_list

DanPuzzuoli committed:
getting first constructor tests to pass

DanPuzzuoli committed:
adding sparse list tests

DanPuzzuoli committed:
re-adding more tests

DanPuzzuoli committed:
pickling working

DanPuzzuoli committed:
saving

DanPuzzuoli committed:
reintroducing apply_layout tests

DanPuzzuoli committed:
final test added

DanPuzzuoli committed:
some formatting

DanPuzzuoli committed:
formatting

ShellyGarion reviewed (commented):
This is a very well written PR, which also includes comprehensive tests. I don't have any comments on the code or on the tests. Perhaps @jakelishman has some comments or suggestions.
I would only like to ask if the new classes should be user-facing? as I don't see them in the docs, and this PR doesn't have any release notes.

DanPuzzuoli committed:
adding release note

DanPuzzuoli committed:
add classes to auto summary for quantum_info module

DanPuzzuoli commented:
Hi @ShellyGarion - I've just added a release note and added `PhasedQubitSparsePauli` and `PhasedQubitSparsePauliList` to the class summary for the `quantum_info` module. Thanks for pointing these out!

DanPuzzuoli committed:
fixing docs error

ShellyGarion commented on releasenotes/notes/phased-qubit-sparse-pauli-eb4274866b543538.yaml:
```diff
@@ -0,0 +1,8 @@
+---
+features:
```
should be `features_quantum_info`

DanPuzzuoli replied:
fixed, thanks!

DanPuzzuoli committed:
updating 'features' to 'features_quantum_info' in the release note

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/phased_qubit_sparse_pauli.rs:
```diff
@@ -0,0 +1,1586 @@
+        let phase_str = match (self.phase - num_ys).rem_euclid(4) {
+            0 => "",
+            1 => "(-i)",
+            2 => "(-1)",
```
Are the parentheses necessary?  Feels like that might kind in the way of the reverse parsing, if so.

DanPuzzuoli replied:
Yeah that's a good question...

So it definitely isn't necessary - this method is only used in generating the string representation of either a `PhasedQubitSparsePauli` or a `PhasedQubitSparsePauliList`, and I chose to put the parentheses purely based on how I thought the output text looks. E.g. 
```
<PhasedQubitSparsePauliList with 1 element on 5 qubits: [(-i)Y_4 Y_3 X_2 Y_1 Z_0]>
```
v.s.
```
<PhasedQubitSparsePauliList with 1 element on 5 qubits: [-iY_4 Y_3 X_2 Y_1 Z_0]>
```

As I'm re-looking at it now though I think it looks fine :P .

In general I don't think I'm too concerned about this text needing to be parsed, but that being said I think it looks fine without the parentheses so I'll remove them.

jakelishman replied:
Oh yeah, I forgot this was the whole "sparse repr" thing and not producing a dense string.  In that case it doesn't matter much either way, so let's stick with the current state.

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/phased_qubit_sparse_pauli.rs:
```diff
@@ -0,0 +1,1586 @@
+        let other_indices = other.qubit_sparse_pauli.indices();
+        let other_paulis = other.qubit_sparse_pauli.paulis();
+
+        // iterate through each entry of self and other one time, incrementing based on the ordering
+        // or equality of self_idx and other_idx, until one of them runs out of entries
+        while self_idx < self_indices.len() && other_idx < other_indices.len() {
+            if self_indices[self_idx] < other_indices[other_idx] {
+                // if the current qubit index of self is strictly less than other, append the pauli
+                paulis.push(self_paulis[self_idx]);
+                indices.push(self_indices[self_idx]);
+                self_idx += 1;
+            } else if self_indices[self_idx] == other_indices[other_idx] {
+                // if the indices are the same, perform multiplication and append if non-identity
+                let new_pauli = (self_paulis[self_idx] as u8) ^ (other_paulis[other_idx] as u8);
+                if new_pauli != 0 {
+                    paulis.push(match new_pauli {
+                        0b01 => Ok(Pauli::Z),
+                        0b10 => Ok(Pauli::X),
+                        0b11 => Ok(Pauli::Y),
+                        _ => Err(ArithmeticError::PauliMultiplication { b: new_pauli }),
+                    }?);
+                    indices.push(self_indices[self_idx])
+                }
+                if (self_paulis[self_idx] == Pauli::X || self_paulis[self_idx] == Pauli::Y)
+                    && (other_paulis[other_idx] == Pauli::Z || other_paulis[other_idx] == Pauli::Y)
+                {
+                    new_phase += 2;
+                }
+                self_idx += 1;
+                other_idx += 1;
+            } else {
+                // same as the first if block but with roles of self and other reversed
+                paulis.push(other_paulis[other_idx]);
+                indices.push(other_indices[other_idx]);
+                other_idx += 1;
+            }
+        }
```
We can go to merge with this in place, but `SparseObservable` defines a `PairwiseOrdered` iterator that you could probably use for much of this.

DanPuzzuoli replied:
Yeah I think it will be easier to deal with this later.

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/phased_qubit_sparse_pauli.rs:
```diff
@@ -0,0 +1,1586 @@
+/// A Pauli operator is a tensor product of single-qubit Pauli operators of the form :math:`P =
+/// (-i)^n \bigotimes_n A^{(n)}_i`, for :math:`A^{(n)}_i \in \{I, X, Y, Z\}` and an integer
+/// :math:`n` called the phase exponent. The internal representation of a
+/// :class:`PhasedQubitSparsePauli` stores only the non-identity single-qubit Pauli operators.
```
Seems like you use `n` for more than one thing here.

DanPuzzuoli replied:
Good catch, thanks, fixed.

jakelishman commented on test/python/quantum_info/test_qubit_sparse_pauli.py:
```diff
@@ -873,38 +873,6 @@ def test_len(self):
         self.assertEqual(len(QubitSparsePauliList.empty(10)), 0)
         self.assertEqual(len(QubitSparsePauliList.from_list(["IIIXIZ", "YYXXII"])), 2)
 
-    def test_pauli_enum(self):
-        # These are very explicit tests that effectively just duplicate magic numbers, but the point
-        # is that those magic numbers are required to be constant as their values are part of the
-        # public interface.
-
```
Why is this test deleted?

DanPuzzuoli replied:
The enum is accessible through both `QubitSparsePauli.Pauli` and `QubitSparsePauliList.Pauli`, so I ended up removing this test because it is redundant with the test method of the same name in `TestQubitSparsePauli`. The only thing additionally that this would be testing is that `QubitSparsePauliList.Pauli` exists, but I believe it's used throughout the tests anyway. I'm happy to add it back or to add a greatly reduced version of it, but it seemed okay to remove at the time.

jakelishman reviewed (dismissed):
I can't say I did the most detailed review ever, but if you're fine with this, you've been using it, and Shelly is fine with it, I'm ok too, bar the couple of very minor comments, which can be follow-ups anyway.

I'll tag for merge now, but maybe we can squash them in #14846

DanPuzzuoli committed:
removing parentheses from str representation

DanPuzzuoli committed:
fixing doc string collision of variable 'n'

DanPuzzuoli committed:
Merge branch 'main' into phased-qubit-sparse-pauli

jakelishman committed:
Remove unnecessary cast
